### PR TITLE
Fix "tip not cached" anomaly when there are open proofs at file end

### LINKED ISSUE #16335: Anomaly "Stm.join: tip not cached"
#### Description of the problem

```coq
Goal True.
Proof.
```
gives
```
Error: Anomaly "Stm.join: tip not cached"
Please report at http://coq.inria.fr/bugs/.
```

cc @coq/stm-maintainers 

#### Coq Version

https://gitlab.com/coq/coq/-/jobs/2740357137/artifacts/download i.e. a17855e0870d17cd722adfd96b867be2de219445

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #16335

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: toplevel/ccompile.ml [hunk 1]
```diff
@@ -68,9 +68,10 @@ let compile opts stm_options injections copts ~echo ~f_in ~f_out =
       let check = Stm.AsyncOpts.(stm_options.async_proofs_mode = APoff) in
       let source = source copts ldir long_f_dot_in in
       let state = Vernac.load_vernac ~echo ~check ~interactive:false ~state ~source long_f_dot_in in
-      let state = Stm.join ~doc:state.doc in
+      let fullstate = Stm.finish ~doc:state.doc in
+      ensure_no_pending_proofs ~filename:long_f_dot_in fullstate;
+      let () = Stm.join ~doc:state.doc in
```
Comment (Alizter) [inline toplevel/ccompile.ml:73]:
```suggestion
      Stm.join ~doc:state.doc;
```

**Review by Alizter — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
This PR is difficult to backport as it conflicts (in particular) with #17059. I would appreciate it if someone takes care of preparing a backport.

**Comment by SkySkimmer** [comment]
https://github.com/coq/coq/pull/17750

**Comment by SkySkimmer** [comment]
Why postpone this?

**Comment by Zimmi48** [comment]
It's not. Your backport PR will be merged in 8.17, so this PR won't be backported using the script.

**Comment by SkySkimmer** [comment]
So when the backport is merged this PR will get re-milestoned?

**Comment by SkySkimmer** [comment]
Maybe I misunderstood what the PR milestone is supposed to mean

**Comment by Zimmi48** [comment]
It's tracking where this specific PR ended up first. In case there is a different PR with a backport, it is the latter that gets the milestone marking where it was merged. I agree that this edge case is maybe not handled as well as it could be.

**Pull request merged**